src/librustc_metadata/rmeta/decoder.rs

```diff
@@ -1366,25 +1368,25 @@ impl<'a, 'tcx> CrateMetadataRef<'a> {
 
     // This replicates some of the logic of the crate-local `is_const_fn_raw` query, because we
     // don't serialize constness for tuple variant and tuple struct constructors.
-    fn is_const_fn_raw(&self, id: DefIndex) -> bool {
+    fn is_const_fn_raw(&self, id: DefIndex, sess: &Session) -> bool {
         let constness = match self.kind(id) {
-            EntryKind::AssocFn(data) => data.decode(self).fn_data.constness,
-            EntryKind::Fn(data) => data.decode(self).constness,
+            EntryKind::AssocFn(data) => data.decode((self, sess)).fn_data.constness,
+            EntryKind::Fn(data) => data.decode((self, sess)).constness,
             // Some intrinsics can be const fn. While we could recompute this (at least until we
             // stop having hardcoded whitelists and move to stability attributes), it seems cleaner
             // to treat all const fns equally.
-            EntryKind::ForeignFn(data) => data.decode(self).constness,
+            EntryKind::ForeignFn(data) => data.decode((self, sess)).constness,
             EntryKind::Variant(..) | EntryKind::Struct(..) => hir::Constness::Const,
             _ => hir::Constness::NotConst,
         };
         constness == hir::Constness::Const
     }
 
-    fn asyncness(&self, id: DefIndex) -> hir::IsAsync {
+    fn asyncness(&self, id: DefIndex, sess: &Session) -> hir::IsAsync {
```

Comment:
This is a cost that affects all functions, but benefits only a specific diagnostic in an erroneous case.

Reply:
The errors could use `guess_head_span` from the `fn main`'s item span instead.

Reply:
So I think you are suggesting not adding `span` to `hir::IsAsync` but instead using `guess_head_span` in the check in `lib_typchk/lib.rs`?

Reply:
That has the effect that this error will point to the whole line, not just the word `async`. Do we really want this?

Reply:
Yes, pointing to `async fn main` is almost as good as pointing to `async` and doesn't have the metadata encoding/decoding cost in the common case.